File: project1/go_ver/encrypt.go
package main

import (
	"crypto/des"
)

func shuffle(inp []byte, sbox []int) []byte {
	if len(inp) != 4 {
		panic("Mismatched block size!")
	}
	in0 := int(inp[0])
	in1 := int(inp[1])
	in2 := int(inp[2])
	in3 := int(inp[3])

	out0 := (in0 + in1 + in2 + in3) % 256
	out1 := (in0 + in1 + in2) % 256
	out2 := (in0 + in1) % 256
	out3 := in0

	out0 = sbox[out0]
	out1 = sbox[out1]
	out2 = sbox[out2]
	out3 = sbox[out3]

	out := []byte{byte(out0), byte(out1), byte(out2), byte(out3)}
	return out
}

func feistel_round(block []byte, sbox []int) []byte {
	if len(block) != EDES_BLOCK_SIZE {
		panic("Mismatched block size!")
	}
	left := block[:4]
	tmp := left
	right := block[4:]

	outp := shuffle(right, sbox)
	left = right
	right = make([]byte, 0)
	for i := 0; i < len(outp); i++ {
		right = append(right, outp[i]^tmp[i])
	}

	return append(left, right...)
}

func encrypt(password []byte, input_bytes []byte, print_to_stdout bool) []byte {

	key := keygen(password, EDES_KEY_SIZE)
	sboxes := get_sboxes(key, print_to_stdout)
	input_blocks := break_to_blocks(input_bytes)

	input := pad(input_blocks)

	encrypted := make([]byte, 0)

	for tmp := 0; tmp < 16; tmp++ {

		for i := 0; i < len(input); i++ {
			input_copy := make([]byte, len(input[i]))
			copy(input_copy, input[i])
			input[i] = feistel_round(input_copy, sboxes[tmp])

		}
	}

	for _, b := range input {
		encrypted = append(encrypted, b...)
	}
	return encrypted
}

// Encrypt using single DES (not EDES)

func des_encrypt(key []byte, input_bytes []byte) []byte {
	key = keygen(key, 8)
	cipher, err := des.NewCipher(key)
	if err != nil {
		panic(err)
	}

	ciphertext := []byte{}
	input_blocks := break_to_blocks(input_bytes)
	input_blocks = des_pad(input_blocks)

	for {
		if len(input_blocks) > 1 {
			block := make([]byte, 8)
			copy(block, input_blocks[0])

			input_blocks = input_blocks[1:]
			tmp := make([]byte, 8)

			cipher.Encrypt(tmp, block)
			ciphertext = append(ciphertext, tmp...)

		} else if len(input_blocks) == 1 {
			txt := make([]byte, 8)
			copy(txt, input_blocks[0])

			tmp := make([]byte, 8)
			cipher.Encrypt(tmp, txt)
			ciphertext = append(ciphertext, tmp...)
			break
		}

	}
	return ciphertext
}

File: project1/python_ver/encrypt.py
import sys

from keygen import keygen
from constants import EDES_BLOCK_SIZE, ROUND_NUM, EDES_KEY_SIZE,DES_BLOCK_SIZE 
from utils import get_sboxes, pad, read_from_stdin
from Crypto.Cipher import DES
from Crypto.Util.Padding import pad as des_pad


def shuffle(inp: bytes, sbox: list) -> bytes:
    if len(inp) != 4:
        print("Mismatched block size!")
        exit(1)
    in0 = inp[0]
    in1 = inp[1]
    in2 = inp[2]
    in3 = inp[3]
    out0 = (in0+in1+in2+in3) % 256
    out1 = (in0+in1+in2) % 256
    out2 = (in0+in1) % 256
    out3 = in0

    out0 = sbox[out0]
    out1 = sbox[out1]
    out2 = sbox[out2]
    out3 = sbox[out3]
    
    out = out0+out1+out2+out3
    return out


def feistel_round(block: bytes, sbox: list) -> bytes:
    if len(block) != EDES_BLOCK_SIZE:
        print("Mismatched block size!")
        exit(1)
    left = block[:4]
    tmp = left
    right = block[4:8]

    outp = shuffle(right, sbox)
    left = right
    right = bytes([a ^ b for a, b in zip(outp, tmp)])

    return left+right


def encrypt(password: bytes, input_bytes: bytes,print_to_stdout=True) -> bytes:
    """Given a password, use EDES to encrypt data in byte format"""
    # Length is in bytes, 32 bytes -> 256 bits
    key = keygen(password, EDES_KEY_SIZE)
    sboxes = get_sboxes(key,print_to_stdout)
    input_blocks = [input_bytes[n:n+EDES_BLOCK_SIZE]
                    for n in range(0, len(input_bytes), EDES_BLOCK_SIZE)]

    # pad the input blocks (only last block is affected) (or a new block is added!)
    input_blocks = pad(input_blocks)

    ciphertext = b""
    for tmp in range(ROUND_NUM):
        for i in range(len(input_blocks)):
            input_blocks[i] = feistel_round(input_blocks[i], sboxes[tmp])

    for block in input_blocks:
        ciphertext += block
    return ciphertext


def des_encrypt(key: bytes, input_bytes: bytes) -> bytes:
    key = keygen(key, DES_BLOCK_SIZE)
    cipher = DES.new(key, DES.MODE_ECB)
    ciphertext=b""
    input_bytes = [input_bytes[n:n+DES_BLOCK_SIZE] for n in range(0, len(input_bytes),DES_BLOCK_SIZE) ]

    while True:
        if len(input_bytes)==1:
            ciphertext+=cipher.encrypt(des_pad(input_bytes[0],DES_BLOCK_SIZE))
            break
        else:
            ciphertext+=cipher.encrypt(input_bytes[0])
        input_bytes=input_bytes[1:]
    return ciphertext

    

if __name__ == "__main__":
    if len(sys.argv) < 1 or len(sys.argv)>3:
        print("Usage: <content_to_encrypt> | python3 encrypt.py <text_key>")
        print("If you wish to encrypt with DES, append -des to the previous command.")
        exit(1)
    print("[-] Encrypting..")
    key = sys.argv[1].encode("utf-8")
    input_text = read_from_stdin()  
    input_bytes = input_text.encode('utf-8')
    print([i for i in input_bytes])
    
    if len(sys.argv) > 2 and sys.argv[2] == "-des":
        encrypted = des_encrypt(key, input_bytes)
    else:
        encrypted = encrypt(key, input_bytes)
    print("[-] Encrypted!")
    print("[-] Output stored in file 'encrypted'")
    print(encrypted.hex())
    with open("encrypted", "w+") as f:
        f.write(encrypted.hex())


